python/mpi_cmake_modules/documentation_builder.py

```diff
@@ -0,0 +1,529 @@
+        project_source_dir (str): Path to the source file of the project.
+    """
+    # Get project_name
+    project_name = Path(project_source_dir).stem
+
+    # Get the doxygen executable.
+    doxygen = _find_doxygen()
+
+    # Get the resources path.
+    resource_path = _resource_path(project_source_dir)
+
+    # get the Doxyfile.in file
+    doxyfile_in = resource_path / "sphinx" / "doxygen" / "Doxyfile.in"
+    assert doxyfile_in.is_file()
+
+    # Which files are going to be parsed.
+    doxygen_file_patterns = "*.h *.hpp *.hh *.cpp *.c *.cc *.hxx"
+
+    # Where to put the doxygen output.
+    doxygen_output = Path(doc_build_dir) / "doxygen"
+
+    # Parse the Doxyfile.in and replace the value between '@'
+    with open(doxyfile_in, "rt") as f:
+        doxyfile_out_text = (
+            f.read()
+            .replace("@PROJECT_NAME@", project_name)
+            .replace("@PROJECT_SOURCE_DIR@", project_source_dir)
+            .replace("@DOXYGEN_FILE_PATTERNS@", doxygen_file_patterns)
+            .replace("@DOXYGEN_OUTPUT@", str(doxygen_output))
+        )
+    doxyfile_out = Path(doxygen_output) / "Doxyfile"
+    doxyfile_out.parent.mkdir(parents=True, exist_ok=True)
+    with open(str(doxyfile_out), "wt") as f:
+        f.write(doxyfile_out_text)
+
+    bashCommand = doxygen + " " + str(doxyfile_out)
+    process = subprocess.Popen(
+        bashCommand.split(), stdout=subprocess.PIPE, cwd=str(doxygen_output)
+    )
+    output, error = process.communicate()
+    print("Doxygen output:\n", output.decode("UTF-8"))
+    print("Doxygen error:\n", error)
+    print("")
+
+
+def _build_breath_api_doc(doc_build_dir):
+    """
+    Use breathe_apidoc to parse the xml output from Doxygen and generate
+    '.rst' files.
+
+    Args:
+        doc_build_dir (str): Path where to create the temporary output.
+    """
+    breathe_apidoc = _find_breathe_apidoc()
+    breathe_input = Path(doc_build_dir) / "doxygen" / "xml"
+    breathe_output = Path(doc_build_dir) / "breathe_apidoc"
+    breathe_option = "-f -g class,interface,struct,union,file,namespace,group"
+
+    bashCommand = (
+        breathe_apidoc
+        + " -o "
+        + str(breathe_output)
+        + " "
+        + str(breathe_input)
+        + " "
+        + str(breathe_option)
+    )
+    process = subprocess.Popen(
+        bashCommand.split(), stdout=subprocess.PIPE, cwd=str(doc_build_dir)
+    )
+    output, error = process.communicate()
+    print("breathe-apidoc output:\n", output.decode("UTF-8"))
+    print("breathe-apidoc error:\n", error)
+    print("")
+
+
+def _build_sphinx_api_doc(doc_build_dir, project_source_dir):
+    """
+    Use sphinx_apidoc to parse the python files output from Doxygen and
+    generate '.rst' files.
+
+    Args:
+        doc_build_dir (str): Path where to create the temporary output.
+        project_source_dir (str): Path to the source file of the project.
+    """
+    # define input folder
+    project_name = Path(project_source_dir).stem
```

Comment:
Using `name` instead of `stem` here (and in similar calls in other places of this file) might be a bit safer as it should also work in case we ever have a package with a dot in it's name for some reason.

Reply:
ok